rules/run_binary.bzl

```diff
@@ -0,0 +1,105 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+run_binary() build rule implementation.
+
+Runs a binary as a build action. This rule does not require Bash (unlike native.genrule()).
+"""
+
+load("//lib:dicts.bzl", "dicts")
+
+def _make_args(ctx):
+    targets = [ctx.attr.tool]  #+ ctx.attr.srcs
+    resolved = [
+        ctx.expand_location(a, targets) if "$(location" in a else a
+        for a in ctx.attr.args
+    ]
+    result = []
+    for v in resolved:
+        result += ctx.tokenize(v)
```

Comment:
https://github.com/bazelbuild/bazel/blob/616481a6b4ffa1818c04bedefe138880b45fe7fa/src/main/java/com/google/devtools/build/lib/skylarkbuildapi/SkylarkRuleContextApi.java#L348

Reply:
Oh, that's an undocumented method. That's why I couldn't find it.

Is it really useful to call it here? You want to tokenize the result of expand_location?
But this might surprise users who didn't use expand_location? (as in "why did my argument got split?", "do I need to quote it?")

Reply:
> Is it really useful to call it here?

Yes, other `*_binary.args` rules do the same.

> You want to tokenize the result of expand_location?

Yes.

> But this might surprise users who didn't use expand_location? (as in "why did my argument got split?", "do I need to quote it?")

Indeed, if they haven't used `args` in the native `*_binary` rules before.
If they have, the current behavior won't surprise them -- it's consistent with e.g. `cc_binary.args`.

Reply:
Do you have an example where we want to tokenize the output of `expand_location`? If the path has a space, we're going to pass two arguments to the command? Sounds like a bug.

Reply:
I don't.

I know we cannot tokenize before `expand_location`, because `$(location foo)` would be split up and `expand_location` would fail.

To prevent splitting, the user has to pass `args = ["'$(location foo)'"]` (notice the single quotes), so even if the expanded path has a space, it stays intact.

Labels cannot have spaces so most paths don't have space in them, so most users get away with `args=["$(location foo)"]` (no single quotes) most of the time.

Let me repeat, this is consistent with all the other `*_binary.args`  -- not saying they are intuitive, but once you know they are Bash-tokenized (which is documented) then the behavior makes sense.

Reply:
Sorry for the delay. It's important to get the design right (because it will be hard to change it once it gets adoption).

The issues are:
 - `ctx.tokenize` shouldn't be used (as it's not documented).
 - I'm not certain tokenizing items in the list is the right choice.
 - You mentioned consistency with other `_binary.args`, but [documentation says](https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes-binaries) they are subject to Make variable substitution. Should we do it here too?

I think the rule `run_binary` can eventually get a lot of adoption, similar to genrule. So I'd prefer err on the side of caution. Do you mind bringing this to bazel-dev@ for wider discussion?

cc @c-parsons

Reply:
Fair arguments.

Make variable support feels too much. Let's maybe wait until there's demand for it.
On second thought, even Bash tokenization may be too much. The semantics are surely surprising (`["'$(location foo)'"]`), perhaps simplicity trumps expressiveness here. Finally, it's easier to add than to remove features.

I'm happy to remove support for tokenization and keep only `$(location)` expansion.

WDYT? Shall I still bring it up on bazel-dev@?

Reply:
If you remove the tokenize, I'm happy to approve the PR (in my opinion, tokenize was needed only because of make variables).

I agree to keep this rule simple.

Reply:
Done.